src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt

```diff
@@ -230,6 +232,83 @@ internal val Class<*>.allDeclaredFieldWithSuperclasses get(): List<Field> {
     return fields
 }
 
+/**
+ * Traverses all fields of a given object in bfs order.
+ *
+ * In case [treatArrayElementsAsFields] is
+ * set to `false`, then array fields will be traversed instead of elements.
+ *
+ * The [callback] is called for each tuple of `(fieldOwner, field, fieldValue)` of the [obj],
+ * the return value of the callback determines, what object should be traversed further
+ * (it can be something different from `fieldValue`).
+ *
+ * Parameters of [callback]:
+ * - `fieldOwner`: either `Any` - some class instance that owns the field, or `Class<*>` is case if `field` is static.
+ * - `field`: the field that currently is traversed.
+ * - `fieldValue`: the value that was read from the field.
+ */
+internal fun traverseObjectHierarchy(obj: Any, treatArrayElementsAsFields: Boolean = true, callback: (Any, Field, Any?) -> Any?) {
+    val queue = ArrayDeque<Any>()
+    val visitedObjects = Collections.newSetFromMap<Any>(IdentityHashMap())
+
+    queue.add(obj)
+    visitedObjects.add(obj)
+
+    while (queue.isNotEmpty()) {
+        val currentObj = queue.removeFirst()
+        val fields =
+            if (currentObj is Class<*>) emptyList<Field>()
+            else if (currentObj.javaClass.isArray && treatArrayElementsAsFields) {
+                // We do not traverse the actual fields of an array,
+                // instead the elements of the array are treated as its field.
+                // But note that primitive arrays are skipped completely.
+                if (currentObj is Array<*>) {
+                    currentObj.forEach { element ->
+                        element?.let { queue.add(it) }
+                    }
+                }
+                emptyList<Field>()
+            }
+            else if (
+                (currentObj is AtomicReferenceArray<*> || currentObj is kotlinx.atomicfu.AtomicArray<*>) &&
+                treatArrayElementsAsFields
+            ) {
+                // Do the same as before, but for atomic arrays (with non-primitive types)
+                val getElementAt = currentObj.javaClass.getMethod("get", Int::class.javaPrimitiveType)
```

Comment:
Instead of using reflection, you can just cast to `AtomicReferenceArray<*>` or `kotlinx.atomicfu.AtomicArray<*>` type, and use the required methods directly.

Reply:
I made it so for atomic arrays, but for atomicfu array I keep getting errors of type `org.jetbrains.kotlinx.lincheck.UtilsKt: standalone invocation of kotlinx.atomicfu.AtomicArray::getSize that was not traced to previous field load` when I access `size` field. When I use `get(...)` method I get the same kind of error. This seems to be related to [KT-47749](https://youtrack.jetbrains.com/issue/KT-47749) which is marked as fixed. Right now I don't understand why it happens in our case